// bookshelf/test/integration.js
var _        = require('lodash');

module.exports = function(Bookshelf) {

  var Knex     = require('knex');
  var Mongo    = require('mongo-gyro');
  var config   = require(process.env.BOOKSHELF_TEST || './integration/helpers/config');
  var Promise  = global.testPromise;

  var MySQL = Bookshelf.initialize({
    client: 'mysql',
    connection: config.mysql,
    pool: {
      afterCreate: function(connection, callback) {
        return Promise.promisify(connection.query, connection)("SET sql_mode='TRADITIONAL';", []).then(function() {
          callback(null, connection);
        });
      }
    }
  });

  var PostgreSQL = Bookshelf.initialize({
    client: 'postgres',
    connection: config.postgres,
  });

  var SQLite3 = Bookshelf.initialize({
    client: 'sqlite3',
    connection: config.sqlite3,
  });

  var knexSqlite3 = Knex.initialize({
    client: 'sqlite3',
    connection: config.sqlite3
  });

  it('should allow creating a Bookshelf client from a Knex instance', function() {
    var bookshelf = Bookshelf.initialize(knexSqlite3);
    expect(bookshelf.knex).to.equal(knexSqlite3);
  });

  it('should allow creating a new Bookshelf instance with "new"', function() {
    var bookshelf = new Bookshelf(knexSqlite3);
    expect(bookshelf.knex).to.equal(knexSqlite3);
  });

  _.each([MySQL, PostgreSQL, SQLite3], function(bookshelf) {

    describe('Dialect: ' + bookshelf.knex.client.dialect, function() {

      before(function() {
        return require('./integration/helpers/migration')(bookshelf).then(function() {
           return require('./integration/helpers/inserts')(bookshelf);
        });
      });

      this.dialect = bookshelf.knex.client.dialect;

      // Only testing this against mysql for now, just so the toString is reliable...
      if (bookshelf.knex.client.dialect === 'mysql') {
        require('./integration/sql/relation')(bookshelf);
      }

      require('./integration/sql/model')(bookshelf);
      require('./integration/sql/collection')(bookshelf);
      require('./integration/sql/relations')(bookshelf);
      require('./integration/plugins')(bookshelf);
    });

  });

  describe("MongoDB", function() {
    var MongoDB = Bookshelf.initialize({
      client: 'mongodb',
      url: config.mongodb
    });


    before(function() {
      return require('./integration/helpers/mongo')(MongoDB);
    });

    it("should allow creating a Bookshelf client from a Mongo instance", function() {
      var mongo = new Mongo(config.mongodb);
      var bookshelf = Bookshelf.initialize(mongo);
      expect(bookshelf.db).to.equal(mongo);
      expect(bookshelf.mongo).to.equal(mongo);
    });

    it("should allow creating a Bookshelf client from a mongo Config", function() {
      expect(MongoDB.db.url).to.equal(config.mongodb);
    });
  });

  describe("MongoDB Dialect", function() {
    var MongoDB = Bookshelf.initialize({
      client: 'mongodb',
      url: config.mongodb
    });
    
    require('./integration/mongo/model')(MongoDB);
    require('./integration/mongo/collection')(MongoDB);
  });

};
